Make legacy skins animatable

jorolf opened the pull request:
Prereqs:
- [x] https://github.com/ppy/osu-framework/pull/2738

This implements general handling for animatable components, however most components haven't been marked as animatable yet.

Closes #4627

jorolf committed:
make legacy skins animatable

iiSaLMaN commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -52,22 +53,29 @@ protected override void Dispose(bool isDisposing)
         public override Drawable GetDrawableComponent(string componentName)
         {
+            bool animatable = false;
+            (bool looping, double frametime) animationData = (false, 1000 / 60d);
```
why not constant for frametime? (`default_animation_frametime`)

peppy replied:
and why a tuple..?

iiSaLMaN commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -81,15 +89,31 @@ public override Drawable GetDrawableComponent(string componentName)
+                    texture = GetTexture($"{componentName}-{i}");
+                }
+
+                animation.Repeat = animationData.looping;
```
is there a looping animation in skins? if not, this line can be removed

peppy replied:
i believe looping is default behaviour

jorolf replied:
Yes, there are definitely looping animations in stable (e.g. animated back buttons)

peppy replied:
Also slider ball and basically every skinnable element, so I agree it should probably be on by default.

iiSaLMaN commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -81,15 +89,31 @@ public override Drawable GetDrawableComponent(string componentName)
+                if (texture == null)
+                    return null;
+
+                return new Sprite { Texture = texture };
```
```suggestion
                return texture != null ? new Sprite { Texture = GetTexture(componentName) } : null;
```

peppy replied:
this is important. without returning actual `null` the skin fallback logic will not work

jorolf replied:
But I am returning ``null`` a few lines above this. His suggestion is pure syntax sugar...

iiSaLMaN commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -52,22 +53,29 @@ protected override void Dispose(bool isDisposing)
 
         public override Drawable GetDrawableComponent(string componentName)
         {
+            bool animatable = false;
```
seems to be true on most components, would be easier to set it true on default and change its value on non-animatable components (`Play/osu/number-text` in this case)

jorolf replied:
This method is called for more components than the switch specifies (e.g. cursor or background) which shouldn't be animated

smoogipoo commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -81,15 +89,31 @@ public override Drawable GetDrawableComponent(string componentName)
+            {
+                var animation = new TextureAnimation { DefaultFrameLength = animationData.frametime };
+
+                for (int i = 1; texture != null; i++)
```
What about the 0-th frame? Shouldn't it be included?

jorolf replied:
``texture`` already contains the 0-th frame from the first animatable check. If you look at the loop body you can see that loading and adding is reversed exactly because I still have it in that variable.

jorolf committed:
fix some code quality

smoogipoo commented:
Needs to be updated to work with the current master + testing (see: https://github.com/ppy/osu/pull/5497)

```
Unhandled Exception: [runtime:error] 2019-07-30 01:43:14: An unhandled error has occurred.
[runtime:error] 2019-07-30 01:43:14: System.InvalidOperationException: No axis can be relatively sized and automatically sized at the same time.
[runtime:error] 2019-07-30 01:43:14: at osu.Framework.Graphics.Containers.CompositeDrawable.set_RelativeSizeAxes(Axes value) in C:\projects\osu-framework-a4n7e\osu.Framework\Graphics\Containers\CompositeDrawable.cs:line 1522
[runtime:error] 2019-07-30 01:43:14: at osu.Game.Skinning.SkinnableDrawable.Update() in /Users/smoogipoo/Repos/osu/osu.Game/Skinning/SkinnableDrawable.cs:line 105
[runtime:error] 2019-07-30 01:43:14: at osu.Framework.Graphics.Drawable.UpdateSubTree() in C:\projects\osu-framework-a4n7e\osu.Framework\Graphics\Drawable.cs:line 443
```

jorolf committed:
Merge branch 'master' of git://github.com/ppy/osu into animated-legacy-skins

jorolf committed:
fix crash and add some tests

- still missing special skin textures

EVAST9919 commented:
Imo every skinnable object in the future should be skinnable

smoogipoo commented:
No animation tests?

jorolf commented:
The hit result animation runs at 60fps which means that for a 1 second animation I would need 60 full size png files. Not sure if we want that, the other solution would be a static or unnoticeable animation.

smoogipoo commented:
The reason I asked for tests is because this doesn't actually work in the case of an animation. The base `Animation` class is a very old class that needs to be re-implemented with proper sizing support.

jorolf commented:
Nice catch, should I update the ``Animation`` class or should I implement a local workaround?

jorolf committed:
add special skin hit "animations"

smoogipoo commented:
I've changed the way animation resizes, but haven't PRd it. Wait a little while so I can see if it's the correct direction.

smoogipoo committed:
Remove explicit autosize

smoogipoo commented:
@peppy is the way you'd see the metrics skin being implemented?

![image](https://user-images.githubusercontent.com/1329837/62758798-656f7f00-baba-11e9-9ce7-852d0728e37a.png)

peppy commented:
Is one of them animated?

smoogipoo commented:
Not yet, no, but I don't know how the metrics skin should look.

peppy committed:
Merge branch 'master' into animated-legacy-skins

peppy committed:
Merge branch 'master' into animated-legacy-skins

peppy committed:
Add sample usage of looping skinnable

peppy committed:
Rename const and make more available

peppy committed:
Avoid unnecessary texture retrieval for non-animated sprites

peppy committed:
Refactor texture lookup code

peppy committed:
Add actual testing of animations

peppy committed:
Further improve legibility of texture lookup

peppy commented:
@smoogipoo i'm done with this. need a re-re-re-view.

peppy committed:
Merge branch 'master' into animated-legacy-skins

peppy committed:
Fix file extension case #1

peppy committed:
Fix file extension case #2

peppy committed:
Don't show judgement type "none"

peppy committed:
Avoid test failures

This is a bit of a workaround (I'm pretty sure these tests are not running correctly on appveyor) but I think it will do for now. Cannot repro failures locally at all.

smoogipoo committed:
Merge branch 'master' into animated-legacy-skins
